src/utilities.jl

```diff
@@ -0,0 +1,89 @@
+using MCMCChains, StatsFuns
+export pointwise_loglikes, compute_loo
+
+"""
+    pointwise_loglikes(chain::Chains, data, ll_fun)
+
+Computes the pointwise log likelihoods where [d,s,c] corresponds to log likelihood of 
+evaluated at datapoint d, sample s, for chain c. 
+
+- `chain`: a chain object from MCMCChains
+- `data`: a vector of data used to estimate parameters of the model 
+- `ll_fun`: a function that computes the log likelihood of a single data point: f(θ1, ..., θn, data), where θi is the ith parameter
+"""
+function pointwise_loglikes(chain::Chains, data, ll_fun)
+    samples = Array(Chains(chain, :parameters).value)
+    pointwise_loglikes(samples, data, ll_fun)
+end
+
+"""
+    pointwise_loglikes(samples::Array{Float64,3}, data, ll_fun)
+
+Computes the pointwise log likelihood 
+
+- `samples`: a three dimensional array of mcmc samples where the first dimension represents the samples, the second dimension represents
+the samples, and the third dimension represents the chains. 
+- `data`: a vector of data used to estimate parameters of the model 
+- `ll_fun`: a function that computes the log likelihood of a single data point: f(θ1, ..., θn, data), where θi is the ith parameter
+"""
+function pointwise_loglikes(samples::Array{Float64,3}, data, ll_fun)
+    fun = (p,d)-> ll_fun(p..., d)
```

Comment:
I believe this introduces a type instability, since the Julia compiler doesn't know the type of the value returned by `ll_fun` until runtime. You can use a parametric type and a "Wrapper" function to enforce type stability here. For instance:
```
function safe_ll_fun(ll_fun::Function, type::Type, args...)
    return dummy::type = ll_fun(args...)
end
```
then, instead of calling `ll_fun`, call `safe_ll_fun(ll_fun, F)`, where `F<:AbstractFloat` is the float type I discussed in my comment above. (This is also why we should try to annotate arguments and return values whenever it's not too difficult -- it lets us catch type instabilities.)

Reply:
I have not seen this pattern before. Interesting. 

In this particular case, I think the function is type stable:

```
@code_warntype fun(samples[s,:,c], data[d])

Variables
  #self#::Core.Const(ParetoSmooth.var"#79#80"{typeof(compute_loglike)}(compute_loglike))
  p::Vector{Float64}
  d::Float64

Body::Float64
1 ─ %1 = Core.getfield(#self#, :ll_fun)::Core.Const(compute_loglike)
│   %2 = Core.tuple(d)::Tuple{Float64}
│   %3 = Core._apply_iterate(Base.iterate, %1, p, %2)::Float64
└──      return %3
```

Reply:
@itsdfish I believe it is in this particular case, but the problem arises if the user provides a function that the compiler can't easily prove won't introduce type instabilities.